### PR TITLE
feat(files): Implement files list filters

### LINKED ISSUE #45437: Filetype filters on top of the files list (doc, pdf, jpg, ...)
@nextcloud/designers for mockups

### PR DESCRIPTION
**Pull request by susnux**
* Resolves: #45437 
* :warning: This requires this (incl. new release) first: https://github.com/nextcloud-libraries/nextcloud-files/pull/1027

## Summary
Current state:

[Bildschirmaufnahme_20240724_201030.webm](https://github.com/user-attachments/assets/7a08781d-bcd4-453a-b233-7aaa6be64aca)

<details>
Old state:

[Screencast_20240607_141239.webm](https://github.com/nextcloud/server/assets/1855448/2bd2da5c-37e0-4420-810b-49265af38eff)

</details>

## TODO

Not related but there are some visual issues, those will be fixed with next nextcloud-vue update (size of the user select etc).

- [x] Implement "people" filter
- [x] Design
- [ ] Tests (follow up)

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by AndyScherzinger — APPROVED** [review]
🐘

File: apps/files/src/eventbus.d.ts [hunk 1]
```diff
@@ -2,14 +2,19 @@
...
 	export interface NextcloudEvents {
 		// mapping of 'event name' => 'event type'
+		'files:config:updated': { key: string, value: unknown }
+
 		'files:favorites:removed': Node
 		'files:favorites:added': Node
 		'files:node:renamed': Node
+
+		'files:filter:add': IFileListFilter
+		'files:filter:remove': string
```
Comment (skjnldsv) [inline apps/files/src/eventbus.d.ts:17]:
```suggestion
		'files:filter:added': IFileListFilter
		'files:filter:removed': string
```

File: apps/files/src/views/FilesList.vue [hunk 2]
```diff
@@ -325,6 +305,25 @@ export default defineComponent({
...
+		dirContents(): Node[] {
+			return (this.currentFolder?._children || [])
+				.map(this.getNode)
+				.filter((node: Node) => !!node)
+		},
+
+		dirContentsFiltered(): Node[] {
+			// mentioned to trigger re-filter on change
+			if (this.filtersStore.filtersChanged) {
+				this.filtersStore.filtersChanged = false
```
Comment (skjnldsv) [inline apps/files/src/views/FilesList.vue:317]: `Unexpected side effect in "dirContentsFiltered" computed property`

File: apps/files/src/store/filters.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		},
+
+		onFilterUpdateChips(event: FilterUpdateChipsEvent) {
+			const id = (event.target as IFileListFilter).id
+			this.chips = { ...this.chips, [id]: [...event.detail] }
+		},
+
+		init() {
+			subscribe('files:filter:add', this.addFilter)
+			subscribe('files:filter:remove', this.removeFilter)
```
Comment (skjnldsv) [inline apps/files/src/store/filters.ts:72]:
```suggestion
			subscribe('files:filter:added', this.addFilter)
			subscribe('files:filter:removed', this.removeFilter)
```

File: apps/files/src/store/filters.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+			logger.debug('New file list filter registered', { id: filter.id })
+		},
+
+		removeFilter(filterId: string) {
+			const index = this.filters.findIndex(({ id }) => id === filterId)
+			if (index > -1) {
+				const [filter] = this.filters.splice(index, 1)
+				filter.removeEventListener('update:chips', this.onFilterUpdateChips)
+				filter.removeEventListener('update:filter', this.onFilterUpdate)
+				logger.debug('Rile list filter unregistered', { id: filterId })
```
Comment (skjnldsv) [inline apps/files/src/store/filters.ts:57]:
```suggestion
				logger.debug('Files list filter unregistered', { id: filterId })
```

File: apps/files/src/views/FilesList.vue (hunk 2 above)
Comment (susnux) [reply apps/files/src/views/FilesList.vue]: fixed

**Review by skjnldsv — APPROVED** [review]
Very very nice work!!! :star_struck:

**Pull request merged**